KadAction

kummy98 opened the pull request.

turcato-niccolo committed:
Merge branch 'feature/smshandler' into feature/networkmanager-Turcato

# Conflicts:
#	lib/src/main/java/com/dezen/riccardo/smshandler/Peer.java
#	lib/src/main/java/com/dezen/riccardo/smshandler/SMSReceiver.java
#	lib/src/main/java/com/dezen/riccardo/smshandler/database/SMSEntity.java

turcato-niccolo committed:
Added separation String handling (might need a better sequence tho).
Fixed some hard wiring.
Replaced Enum and array of strings with static Class Actions

turcato-niccolo committed:
Fixed some leftover of the last commit

turcato-niccolo committed:
Fixed bugs in NetworkDictionaryTest: SMSPeers were created with non valid address
Added getResourceByName and getPeerByAddress in NetworkDictionary

Commit:
Improved interfaces for Netowork manager.

- NetworkInterface is a more generic interface, better fitting any kind of network
- ReplicatedNetworkInterface is meant to describe the behaviour of a replicated network
- ReplicatedNetworkManager is the new name for NetworkManager
- Fixed some comments.

Signed-off-by: Riccardo De Zen <riccardo.dezen@studenti.unipd.it>

turcato-niccolo committed:
Merge branch 'feature/networkmanager' into feature/networkmanager-Turcato

# Conflicts:
#	lib/src/main/java/com/dezen/riccardo/smshandler/SMSReceiver.java
#	networkmanager/src/main/java/com/dezen/riccardo/networkmanager/NetworkManager.java

turcato-niccolo committed:
Added isGlobalPhoneNumber in SMSPeer to replace PhoneNumberUtils.isGlobalPhoneNumber()

Added removeKeysFromMessageBody, for both Strings and SMSMessages in SmsUtils

Added equals in SMSMessage

Added tracking of deleted resources and Peers in NetworkDictionary, for a correct export to backup Database

Fixed some old tests

turcato-niccolo committed:
Fixed equals method in SMSMessage, after Giorgia Bortoletti's review

turcato-niccolo committed:
Added NetworkManagerAction and ActionStructure, class for encapsulate syntax for NetworkManager (and possible other future managers)

turcato-niccolo committed:
Updated SMSHandler, non-priority Messages forwared to listeners are stripped of APP_KEY

Added Support for Actions (abstract ActionStructure and NetworkAction) in NetworkManager

turcato-niccolo committed:
Merge branch 'feature/networkmanager' into feature/networkmanager-Turcato

# Conflicts:
#	networkmanager/src/main/java/com/dezen/riccardo/networkmanager/NetworkManager.java

turcato-niccolo committed:
Updated with latest commit on feature/networkManager: manual merge of class NetworkManager required

bortoletti-giorgia committed:
DictionaryDao now extends a generic BaseDao to remove hardwiring.

bortoletti-giorgia committed:
Updated BaseDao for DictionaryDao

ricdezen committed:
Merge pull request #40 from RicDeZen/feature/networkmanager-Turcato

Added Actions to networkmanager

bortoletti-giorgia committed:
Removed other hardwiring

Remove hardwiring:
- column name in the peer and resource table
- move getPeer and getResource from BaseDao to DictionaryDao because knowing that 'key' and 'address' are String violates the generic types a little bit

Commit:
Merge branch 'feature/distributedNetwork' of https://github.com/RicDeZen/SmsHandler into feature/networkmanager

Signed-off-by: Riccardo De Zen <riccardo.dezen@studenti.unipd.it>

# Conflicts:
#	networkmanager/src/main/java/com/dezen/riccardo/networkmanager/NetworkManager.java

ricdezen committed:
Merge pull request #41 from RicDeZen/feature/networkmanager-basedao

Feature/networkmanager basedao

ricdezen committed:
Merge pull request #43 from RicDeZen/feature/networkmanager-replicated-refactor

Added better interfaces for Networks

turcato-niccolo committed:
-Added abstract class NodeActionStructure<T>

turcato-niccolo committed:
-Updated abstract class NodeActionStructure<T>

Commit:
added class DistributedNetworkAction
Some method still need to be implemented (setDestination getDestination and getSender)

Commit:
-Changed STORE's extra from IGNORED to NAME
-Changed ANSWER_FIND_VALUE's extra from YES_RESOURCES to NAME
-Added totalMassages and current messages to the constructor
-removed the 2 extra check

Commit:
-Removed totalMassages and current messages from the constructor
-added 2 methods to set currentMessage and totalMessage

Commit:
-changed return object from Node to SMSPeer in methods getDestination,
getSender and the parameter of sedDestination from Node to SMSPeer

Commit:
Renamed Classes

- The abstract class/interface should be the "distributed network" action
- The implementing class is KadAction since it uses Kademlia

Signed-off-by: Riccardo De Zen <riccardo.dezen@studenti.unipd.it>

Commit:
Some fixes to both classes

- DistributedNetworkAction is now an interface. Java only allows extending one class, but multiple interfaces, using an abstract class should be the choice when actual implementation of methods or parameters is required. Which is not in this case. Also the functionality has been extremely trimmed down. An Interface for a generic Action won't know what all the various parameters are.
- KadAction now contains two Enums, one for Action type, one for Content type.

Signed-off-by: Riccardo De Zen <riccardo.dezen@studenti.unipd.it>

Commit:
Removed dem empty lines cause of lingering OCD

Signed-off-by: Riccardo De Zen <riccardo.dezen@studenti.unipd.it>

Commit:
Renamed fields in KadAction, removed unnecessary methods.

Signed-off-by: Riccardo De Zen <riccardo.dezen@studenti.unipd.it>

Commit:
Static method for StringResource parsing

Signed-off-by: Riccardo De Zen <riccardo.dezen@studenti.unipd.it>

Commit:
Fixed mistake

Signed-off-by: Riccardo De Zen <riccardo.dezen@studenti.unipd.it>

Commit:
Fixed tests for StringResource

Signed-off-by: Riccardo De Zen <riccardo.dezen@studenti.unipd.it>

ricdezen committed:
Merge pull request #54 from RicDeZen/hotfix/string-resource-parse

Static method for StringResource parsing

Commit:
-Changed syntax of message
-Removed extra and params
-Changed parameters for the constructor
-Now we've enum for PayloadType and payload
-Added methods for the check of the payload type

Commit:
Merge branch 'feature/networkmanager' of https://github.com/RicDeZen/SmsHandler into feature/distributedNetwork-kadActions

Commit:
fixed an error

Commit:
Merge branch 'feature/distributedNetwork' of https://github.com/RicDeZen/SmsHandler into feature/distributedNetwork-kadActions

# Conflicts:
#	distributednetworkmanager/src/main/java/com/gruppo1/distributednetworkmanager/DistributedNetworkAction.java

turcato-niccolo commented on distributednetworkmanager/src/main/java/com/gruppo1/distributednetworkmanager/KadAction.java:
```diff
@@ -0,0 +1,285 @@
+            //= actionType;
+            //TODO
+        }
+        if (isValid()) ;
```
some leftover maybe?

turcato-niccolo commented:
KadAction is hella good, I think that the same methodology should be applied to the actions for the action Types of the replicated network.

ricdezen commented on distributednetworkmanager/src/main/java/com/gruppo1/distributednetworkmanager/KadAction.java:
```diff
@@ -0,0 +1,285 @@
+            case PEER_ADDRESS:
+                return SMSPeer.isAddressValid(payload) == SMSPeer.PhoneNumberValidity.ADDRESS_VALID;
+            case RESOURCE:
+                //metodo statico parse resource
```
missing method call

Commit:
-added specs
-added missing cases of the switch

Commit:
-changed the comparison method from biset.length to bitset.size in the payloadMatchType method

turcato-niccolo commented on distributednetworkmanager/src/main/java/com/gruppo1/distributednetworkmanager/KadAction.java:
```diff
@@ -226,24 +236,41 @@ public boolean isValid() {
             case RESOURCE:
-                //metodo statico parse resource
+                try {
+                    return StringResource.parseString(payload, SEPARATOR).isValid();
```
Wait, are we using the same separator for the parameters AND the resource name:value?
the line `String[] parameteres = messageBody.split(SEPARATOR);` will not work correctly

kummy98 replied:
You're right I should change that SEPARATOR with the separatore used on resources

ricdezen replied:
My original idea was to not even use a separator, since everything is a constant (we know that the number of parts can't go over k and we can define a constant min and max value for operation id, like from 0 to 99). I'd choose this second option instead of using a different separator, we don't know for sure the user won't use our same separator somewhere.
If this solution is used then 0 must be used to fill any voids (i.e. operation id or max parts should be 0X instead of X) and to get the String array with the various parts, substrings should be used

ricdezen replied:
As for StringResource, I made a mistake allowing the method to take a separator, such separator should be known by the Resource class itself.

turcato-niccolo replied:
Well, letting the user define the separator would allow avoiding certain problems (but of course one must be careful and things like this can happen)

ricdezen commented on distributednetworkmanager/src/main/java/com/gruppo1/distributednetworkmanager/KadAction.java:
```diff
@@ -0,0 +1,348 @@
+    public SMSMessage toMessage() {
+        StringBuilder body = new StringBuilder();
+        body.append(actionType.getCode()).append(SEPARATOR).append(operationId)
+                .append(SEPARATOR).append(currentPart).append(SEPARATOR).append(totalParts)
```
Here currentPart, totalParts and operationId should be checked for padding zeros if we're using the option I suggested above

ricdezen commented on distributednetworkmanager/src/main/java/com/gruppo1/distributednetworkmanager/KadAction.java:
```diff
@@ -0,0 +1,348 @@
+            return false;
+        if (this.payloadType.equals(PayloadType.INVALID))
+            return false;
+        if (currentPart <= 0 || totalParts <= 0 || currentPart > totalParts)
```
As suggested below I'd put a static field min = 0 and max = 99, valid for operationId, totalParts and currentParts, so we know it's 2 chars each, if you do so, isValid should check the values are within the bounds

turcato-niccolo replied:
100 is too little in my opinion: if you generate id as random, the probability of colliding with another id generated by another peer is 1%. Let's use at least index of 0-999.

Martignago98 committed:
Added the Class for the processing of the receiving action

ricdezen commented on distributednetworkmanager/src/main/java/com/gruppo1/distributednetworkmanager/DistributedNetworkActionResponder.java:
```diff
@@ -0,0 +1,90 @@
+     */
+    private void sendAction (KadAction responseAction){
+
+        SMSManager.getInstance(context).sendMessage(responseAction.toMessage());
```
If you pass an ActionPropagator in the constructor you don't need to send the message through SMSManager, that will be the ActionPropagator's job (most likely the NetworkManager that will keep everything together). For the same reason you don't really need Context in the constructor, you just need NodeDataProvider, ActionPropagator and maybe some kind of interface for Resource access (the NetworkManager will pass a valid one for each of these when instantiating this class)

ricdezen commented on distributednetworkmanager/src/main/java/com/gruppo1/distributednetworkmanager/DistributedNetworkActionResponder.java:
```diff
@@ -0,0 +1,90 @@
+     * @param action received action
+     */
+    public void onActionReceived(KadAction action) {
+        switch (action.getActionType()) {
```
Here, since you only care about Requests, can add an if to return immediately when you receive a Response (there's a method in the ActionType enum called .isResponse() for that). Remember: you get Requests, you send Responses, that's this class' job

turcato-niccolo commented on distributednetworkmanager/src/main/java/com/gruppo1/distributednetworkmanager/DistributedNetworkActionResponder.java:
```diff
@@ -0,0 +1,90 @@
+    private void onPingReceived(KadAction action) {
+        //NodeDataProvider.iGotCalledByNode(action.getPeer());
+
+        KadAction answerPing =new KadAction(action.getPeer(), KadAction.ActionType.PING_ANSWER,1,1,1, KadAction.PayloadType.BOOLEAN,"true");//TODO inserire id min e max part correti e decidere payload
```
Watch out: the response to a request must have the same id of the req.

Commit:
-Removed the SEPARATOR
-Added 2 methods to remove or add padding
-Added some preCondition to currentPart TotalPars and OperationID
-Added test (I need to rewrite them in a better way)

Commit:
-just a few changes to specs

Commit:
Merge remote-tracking branch 'origin/feature/distributedNetwork-kadActions' into feature/distributedNetwork-kadActions

ricdezen commented on distributednetworkmanager/src/main/java/com/gruppo1/distributednetworkmanager/KadAction.java:
```diff
@@ -0,0 +1,369 @@
+            return false;
+        if (this.payloadType.equals(PayloadType.INVALID))
+            return false;
+        if (currentPart <MIN_PARTS || totalParts <= MIN_PARTS || currentPart > totalParts || totalParts>MAX_PARTS )
```
Sono 35 minuti che non capisco cosa cazzo ha la demo che crasha e crasha e crasha e crasha.
totalParts <= MIN_PARTS rende invalida l'Action? Non posso inviare meno di due messaggi? Dude pliz

turcato-niccolo replied:
it's definitely wrong, total <= 1 is not invalid if the total is actually 1, for example, ping has one message of one (1/1) and must be considered valid.
Should be totalParts < MIN_PARTS

kummy98 replied:
Dopo averlo cambiato a totalParts<Min_parts pensavo proprio di togliere quest'ultimo controllo perché se TotalParts<1 allora automaticamente si rientra nella condizione subito dopo ovvero currentPart>totalParts  essendo la condizione subito prima currentPart <MinParts non soddisfatta ;

Commit:
-Min max id,parts are now public
-Fixed an error on check of totalparts

Commit:
-Improved the tests;now the class is covered by 98%
-Removed a check in isValid method

turcato-niccolo reviewed (approved):
![Screenshot from 2019-12-20 10-02-45](https://user-images.githubusercontent.com/56435959/71243128-e44db180-230f-11ea-9705-31da778f83a9.png)

turcato-niccolo commented on distributednetworkmanager/src/main/java/com/gruppo1/distributednetworkmanager/KadAction.java:
```diff
@@ -0,0 +1,379 @@
+                return true;
+            case NODE_ID:
+                BitSet bitset=BitSetUtils.decodeHexString(payload);
+                return bitset.size()==ID_LENGTH;
```
I'm so very sorry I've seen this just now, you can't do this control, because bitsets aren't of fixed length.
Just check if you can build the bitset (no exceptions) and if size <= ID_LENGTH.
This happens because BitSets only save the positions of bits set to 1.
Btw I updated BinarySet some time ago adding a contructor from hexadecimal string

turcato-niccolo replied:
I noticed it just now by testing the builder

turcato-niccolo commented on distributednetworkmanager/src/main/java/com/gruppo1/distributednetworkmanager/KadAction.java:
```diff
@@ -0,0 +1,379 @@
+    private static final int ID_LENGTH = 128;
+    public static final int MIN_ID = 1;
+    public static final int MAX_ID = 999;
+    private static final String RESOURCE_SEPARATOR="\r";
```
Also this, it should be public, or the asshole outside (being me) can't build resource payload
Anyway we should also add a better parsing strategy

Commit:
Update

Signed-off-by: Riccardo De Zen <riccardo.dezen@studenti.unipd.it>

Commit:
Merge branch 'feature/distributedNetwork' of https://github.com/RicDeZen/SmsHandler into feature/distributedNetwork-kadActions

Signed-off-by: Riccardo De Zen <riccardo.dezen@studenti.unipd.it>

# Conflicts:
#	distributednetworkmanager/src/main/java/com/gruppo1/distributednetworkmanager/DistributedNetworkAction.java
